onnxruntime/core/framework/kernel_registry_manager.cc

```diff
@@ -14,63 +14,105 @@ Status KernelRegistryManager::CreateKernel(const onnxruntime::Node& node,
                                            /*out*/ std::unique_ptr<OpKernel>& op_kernel) const {
-  std::lock_guard<OrtMutex> lock(lock_);
-  if (kernel_registries_.empty()) {
-    return Status(ONNXRUNTIME, FAIL, "Kernel not found.");
+  const std::string& ptype = node.GetExecutionProviderType();
+  if (ptype.empty()) {
+    return Status(ONNXRUNTIME, FAIL, "The node is not placed on any Execution Provider, therefore, can't find a suitable kernel for it");
   }
-
   Status status;
-  for (auto& registry : kernel_registries_) {
-    status = registry->CreateKernel(node,
-                                    execution_provider,
-                                    session_state.GetInitializedTensors(),
-                                    session_state.GetMLValueNameIdxMap(),
-                                    session_state.GetFuncMgr(),
-                                    op_kernel);
+  {
+    std::lock_guard<OrtMutex> lock(lock_);
+    for (auto& registry : kernel_registries_) {
+      status = registry->CreateKernel(node,
+                                      execution_provider,
+                                      session_state.GetInitializedTensors(),
+                                      session_state.GetMLValueNameIdxMap(),
+                                      session_state.GetFuncMgr(),
+                                      op_kernel);
+      if (status.IsOK()) {
+        return status;
+      }
+    }
+  }
+
+  KernelRegistry* p = nullptr;
+  {
+    std::lock_guard<OrtMutex> lock(lock_);
+    auto iter = provider_type_to_registry_.find(ptype);
+    if (iter != provider_type_to_registry_.end()) p = iter->second.get();
+  }
+  if (p != nullptr) {
+    status = p->CreateKernel(node,
+                             execution_provider,
+                             session_state.GetInitializedTensors(),
+                             session_state.GetMLValueNameIdxMap(),
+                             session_state.GetFuncMgr(),
+                             op_kernel);
     if (status.IsOK()) {
       return status;
     }
   }
 
-  return status;
+  std::ostringstream errormsg;
+  errormsg << "Failed to find kernel for " << node.OpType();
+  if (node.Op() != nullptr)
+    errormsg << "(" << node.Op()->since_version() << ")";
+  if (!node.Name().empty()) errormsg << " (node " << node.Name() << ")";
+  return Status(ONNXRUNTIME, FAIL, errormsg.str());
 }
 
-void KernelRegistryManager::RegisterKernels(const ExecutionProviders& execution_providers,
-                                            KernelRegistryPriority priority) {
-  for (auto& provider : execution_providers)
-    RegisterKernelRegistry(provider->GetKernelRegistry(), priority);
+void KernelRegistryManager::RegisterKernels(const ExecutionProviders& execution_providers) {
+  for (auto& provider : execution_providers) {
+    std::lock_guard<OrtMutex> lock(lock_);
+    auto iter = provider_type_to_registry_.find(provider->Type());
```

Comment:
Is there a reason to keep acquiring/freeing the lock within the loop instead of outside it?

Reply:
Very tiny difference. 
Because the code
```c++
for (auto& provider : execution_providers)
```
will try to invoke some functions in ExecutionProviders class. If any of these functions tried to get the lock in KernelRegistryManager again,  it will be a deadlock.
Usually, I would prefer to release a lock as soon as possible to avoid deadlock, but in this case, as you pointed, there is no such need.